### DISCUSSION TITLE
model is moved to device for each new request in fastapi

### DISCUSSION THREAD

**Discussion by girishnadiger-gep** [fastapi/fastapi#6816] [Questions] [answered]
I have a [simpletransformers](https://github.com/ThilinaRajapakse/simpletransformers) based distilbert model, which I am consuming by loading it to memory, then to device (cuda) in the app during runtime and storing it as global variable, so that for new requests I just pass the model for prediction.

Issue is for each new request which I receive, the model is pushed to device(cuda) again, and the model size is big (~700MB) which is adding overhead to each endpoint where this model is being consumed.

Can anyone please help?

python 3.7.5
fastapi 0.60.1
simpletransformers 0.27.2
torch 1.15.0
cuda 11.0

**Comment by raphaelauv** [comment, marked as answer]
This doesn't look like a fastapi issue , close this issue if it's the case.

You can ask for help in stackoverflow or https://gitter.im/tiangolo/fastapi

**Comment by girishnadiger-gep** [comment]
Hi @raphaelauv, okay.

**Comment by girishnadiger-gep** [comment]
Hello Future Readers, the issue was the torch was using multiprocessing for feature generation (input data -> tensors) before passing it to model for prediction, so for multiprocessing multiple workers (no.of cpu - 2) were created, which loaded their own copy of model/tokenizer to device(cuda), for each model.predict, which in this case, for each input request.

Resolution was to disable use_multiprocessing for small inputs. use multiprocessing as dataloader only if there is lot of data on which you needs prediction, and considering the overhead of model/tokenizer load time, multiprocessing is faster...

**Comment by tiangolo** [comment]
Thanks for the help here @raphaelauv :coffee: 

Thanks for reporting back and closing the issue @girishnadiger-gep  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
